nifi-nar-bundles/nifi-standard-services/nifi-distributed-cache-services-bundle/nifi-distributed-cache-client-service/src/main/java/org/apache/nifi/distributed/cache/client/CacheClientHandshakeHandler.java

```diff
@@ -0,0 +1,152 @@
+     *
+     * @param channel           the channel to which this {@link io.netty.channel.ChannelHandler} is bound.
+     * @param versionNegotiator coordinator used to broker the version of the distributed cache protocol with the service
+     */
+    public CacheClientHandshakeHandler(final Channel channel, final VersionNegotiator versionNegotiator) {
+        this.promiseHandshakeComplete = channel.newPromise();
+        this.protocol = new AtomicInteger(PROTOCOL_UNINITIALIZED);
+        this.versionNegotiator = versionNegotiator;
+    }
+
+    /**
+     * API providing client application with visibility into the handshake process.  Distributed cache requests
+     * should not be sent using this {@link Channel} until the handshake is complete.
+     */
+    public void waitHandshakeComplete() {
+        promiseHandshakeComplete.awaitUninterruptibly();
+    }
+
+    @Override
+    public void channelActive(final ChannelHandlerContext ctx) throws IOException {
+        final ByteBuf byteBufMagic = Unpooled.wrappedBuffer(MAGIC_HEADER);
+        ctx.write(byteBufMagic);
+        logger.debug("Magic header written");
+        final int currentVersion = versionNegotiator.getVersion();
+        final ByteBuf byteBufVersion = Unpooled.wrappedBuffer(new OutboundAdapter().write(currentVersion).toBytes());
+        ctx.writeAndFlush(byteBufVersion);
+        logger.debug("Protocol version {} proposed", versionNegotiator.getVersion());
+    }
+
+    @Override
+    public void channelRead(final ChannelHandlerContext ctx, final Object msg) {
+        if (promiseHandshakeComplete.isSuccess()) {
+            ctx.fireChannelRead(msg);
+        } else {
+            final ByteBuf byteBuf = (ByteBuf) msg;
+            try {
+                processHandshake(ctx, byteBuf);
+            } catch (IOException | HandshakeException e) {
+                throw new IllegalStateException("Handshake Processing Failed", e);
+            } finally {
+                byteBuf.release();
+            }
+        }
+    }
+
+    /**
+     * Negotiate distributed cache protocol version with remote service.
+     *
+     * @param ctx     the {@link Channel} context
+     * @param byteBuf the byte stream received from the remote service
+     * @throws HandshakeException on failure to negotiate protocol version
+     * @throws IOException        on write failure
+     */
+    private void processHandshake(final ChannelHandlerContext ctx, final ByteBuf byteBuf) throws HandshakeException, IOException {
+        final short statusCode = byteBuf.readUnsignedByte();
+        if (statusCode == ProtocolHandshake.RESOURCE_OK) {
+            logger.debug("Protocol version {} accepted", versionNegotiator.getVersion());
+            protocol.set(versionNegotiator.getVersion());
+        } else if (statusCode == ProtocolHandshake.DIFFERENT_RESOURCE_VERSION) {
+            final int newVersion = byteBuf.readInt();
+            logger.debug("Protocol version {} counter proposal", newVersion);
```

Comment:
It would be helpful to clarify that this version received in the message:
```suggestion
            logger.debug("Received Protocol version {} counter proposal", newVersion);
```